src/libsync/propagateupload.cpp

```diff
@@ -894,5 +898,392 @@ void PropagateUploadFile::abortWithError(SyncFileItem::Status status, const QStr
+            //done(SyncFileItem::SoftError, tr("Local file changed during sync."));
+            //return;
+        }
+
+        QHttpPart bundleContent;
+        bundleContent.setHeader(QNetworkRequest::ContentLengthHeader, QVariant(fileSize));
+        contentID++;
+        bundleContent.setRawHeader("Content-ID", QByteArray::number(contentID));
+        bundleContent.setRawHeader("X-OC-Mtime", QByteArray::number(qint64(item->_modtime)));
+        if(item->_file.contains(".sys.admin#recall#")) {
+            // This is a file recall triggered by the admin.  Note: the
+            // recall list file created by the admin and downloaded by the
+            // client (.sys.admin#recall#) also falls into this category
+            // (albeit users are not supposed to mess up with it)
+
+            // We use a special tag header so that the server may decide to store this file away in some admin stage area
+            // And not directly in the user's area (which would trigger redownloads etc).
+            bundleContent.setRawHeader("OC-Tag", QByteArray(".sys.admin#recall#"));
+        }
+
+        //TODO use Upload debive to support bandwith limitation on the client
+        QFile *file = new QFile(fullFilePath);
+        if (! file->open(QIODevice::ReadOnly)) {
+            qDebug() << "ERR: Could not prepare upload device";
+
+            // If the file is currently locked, we want to retry the sync
+            // when it becomes available again.
+            if (FileSystem::isFileLocked(fullFilePath)) {
+                emit _propagator->seenLockedFile(fullFilePath);
+            }
+
+            // Soft error because this is likely caused by the user modifying his files while syncing
+            //abortWithError( SyncFileItem::SoftError, "ERR: Could not prepare upload device" );
+            delete file;
+            //return;
+        }
+        bundleContent.setBodyDevice(file);
+        file->setParent(multiPart);
+        multiPart->append(bundleContent);
+    }
+
+    _propagator->_activeJobList.append(this);
+
+    const QString userPath = _propagator->account()->davFilesPath();
+
+    // job takes ownership of device via a QScopedPointer. Job deletes itself when finishing
+    MultipartJob* job = new MultipartJob(_propagator->account(), userPath, multiPart);
+
+    //_jobs.append(job);
+    connect(job, SIGNAL(finishedSignal()), this, SLOT(slotMultipartFinished()));
+    connect(job, SIGNAL(destroyed(QObject*)), this, SLOT(slotJobDestroyed(QObject*)));
+    job->start();
+}
+
+QString PropagateBundleQNAM::getRemotePath(QString filePath){
+    return QString((_propagator->_remoteFolder) + filePath);
+}
+
+bool PropagateBundleQNAM::append(SyncFileItemPtr bundledFile){
+    //TODO: validate here that
+    _bundledFiles.append(bundledFile);
+    QJsonObject fileMetadata;
+    fileMetadata["oc-path"] = getRemotePath(bundledFile->_file);
+    fileMetadata["x-oc-mtime"] = QString::number(qint64(bundledFile->_modtime));
```

Comment:
No, it has to be done here. What you do , you create here the JSON file describing the structure of created bundle.